### PR TITLE
feat(page): Add toggle for full width page view

### PR DESCRIPTION
**Pull request by mejo-**
The full width mode is a per-page user setting and is persisted in browser storage. It will not persist across devices.

Fixes: #242

### Screenshots

| Nomal | Full width |
| --- | --- |
| grafik | grafik |

### 🏁 Checklist

- [x] Code is properly formatted (`npm run lint` / `npm run stylelint` / `composer run cs:check`)
- [x] [Sign-off message](https://probot.github.io/apps/dco/) is added to all commits
- [x] Tests (unit, integration and/or end-to-end) passing and the changes are covered with tests
- [x] Documentation is not required

### CODE REVIEW

**Review by nimishavijay — CHANGES_REQUESTED** [review]
Super nice! Notes from the design review:

- [x] Move "Last changed by" item to the top
- [x] Add divider before and after "Full width" and "Show outline"
```
Jane Doe an hour ago
---
Full width
Show outline 
---
Add template for subpages
Show in Files
Delete page
```

- [ ] Bug: if full-width is toggled, the action menu does not close, but 3 dot menu moves --> to be fixed in component
  - [x] Temporary solution: Always show the group of 3 buttons (Edit, 3dot menu, sidebar toggle) at the right end of the page 
  - [x] Group of buttons should have equal top and right margin
  - [ ] Nice-to-have: buttons are aligned with top right header buttons (avatar, notifications, contact search, universal search)

**Comment by mejo-** [comment]
@nimishavijay I addressed all the topics we discussed in the call, please see the updated screenshots :relaxed:

**Review by nimishavijay — APPROVED** [review]
Super! :)

**Comment by jancborchardt** [comment]
@KaeTuuN @theron29 since you commented in the issue this fixes, would this "per-page per-device" setting work for you?

@mejo- I’m leaning a bit towards that the layout should be a per-device setting across pages, since per the issue it seems it’s mostly people with large screens, and that doesn’t change across pages.

**Comment by juliushaertl** [comment]
> @mejo- I’m leaning a bit towards that the layout should be a per-device setting across pages, since per the issue it seems it’s mostly people with large screens, and that doesn’t change across pages.

From my perspective it is more the actual page content that would influence if you want to use the full width or regular width mode, e.g. for pages with long text you might prefer a limited width for readability while for pages with larger tables or other content a full width is more suitable. I think mid term we should also adapt this in text itself, so having it per document could also be stored in the file rather than per user.

File: cypress/e2e/pages.spec.js [hunk 1]
```diff
@@ -269,6 +269,36 @@ describe('Page', function() {
 		})
 	})
 
+	describe.only('Full width view', function() {
```
Comment (max-nextcloud) [inline cypress/e2e/pages.spec.js:272]:
```suggestion
	describe('Full width view', function() {
```

**Review by max-nextcloud — CHANGES_REQUESTED** [review]
Had a quick look at the code. Looks good all in all.

Just one minor thing that needs to be addressed before the merge.

**Comment by theron29** [comment]
@jancborchardt Yes, this actually looks very good and promising :+1:

File: cypress/e2e/pages.spec.js (hunk 1 above)
Comment (mejo-) [reply cypress/e2e/pages.spec.js]: Good catch, fixed now :relaxed:

**Comment by mejo-** [comment]
Failing Cyress tests are related, will look into it next week.

**Review by marcoambrosini — APPROVED** [review]

**Comment by KaeTuuN** [comment]
@jancborchardt Thanks for mentioning me!

First: Thanks for developing! (@mejo- )
Second:

1. Will this Setting also affect the Edit Mode? (I hope so :-) )
2. Won't it be possible to move the Option directly to the top? Reason: Most complains were about table views and now you have to do more clicks for switching if you just want to view a table on full width and then go back to "normal" mode.

Greetings,
Kae

**Comment by KaeTuuN** [comment]
I'm on my Smartphone atm.
I just want to put it as an Icon next to the edit Button.
Expand Button might look like this: https://cdn4.iconfinder.com/data/icons/ui-controls-editor/32/interface-controls-horizontal-expand-512.png
And the shrink Button might look like this: https://cdn.onlinewebfonts.com/svg/img_278259.png
The Icon changes in case of switching width.

Greetings,
Kae

**Review by jancborchardt — APPROVED** [review]
Thanks for the additional explanation @juliushaertl, then looks good design-wise. :)

And the setting is good in the menu, moving it out would make it too present compared to how important it is.

**Comment by mejo-** [comment]
>     1. Will this Setting also affect the Edit Mode? (I hope so :-) )

Sure :blush: 

>     2. Won't it be possible to move the Option directly to the top? Reason: Most complains were about table views and now you have to do more clicks for switching if you just want to view a table on full width and then go back to "normal" mode.

Well, as @jancborchardt already said, this would be a bit too prominent. We try to not clutter the user interface with too many buttons and switches to not overwhelm users.

File: src/store/pages.js [hunk 2]
```diff
@@ -763,5 +778,37 @@ export default {
...
+		 * @param {object} store.getters getters of the store
+		 * @param {boolean} fullWidthView Whether full width view is enabled or not
+		 */
+		[SET_FULL_WIDTH_VIEW]({ commit, getters }, fullWidthView) {
+			const pageId = getters.currentPage.id
+			const fullWidthPageIds = JSON.parse(persistentStorage.getItem('fullWidthPageIds') ?? '[]')
+			if (fullWidthView && !fullWidthPageIds.includes(pageId)) {
+				fullWidthPageIds.push(pageId)
+				commit(SET_FULL_WIDTH_PAGEIDS, fullWidthPageIds)
+				persistentStorage.setItem('fullWidthPageIds', JSON.stringify(fullWidthPageIds))
```
Comment (juliushaertl) [inline src/store/pages.js:806]: Would be nice to prefix this with the app id, considering we may want to share this state with text at some point maybe we can already use something like `text-fullWidthPageIds`?

**Pull request merged**